Applens Revamp

cindylovescoding opened the pull request:
1. Home page revamp to add category and support topic tiles
2. Left hand side menu revamp
3. Add login user info
4. Add graph API, support topic API, self help content API in applens backend
Home page
![image](https://user-images.githubusercontent.com/38216903/58281357-d1620700-7d57-11e9-8ce4-3291b8af41e4.png)

Detector page: [due to the prod bug right now, detector author info for public detectors is not showing up somehow. Once that get resolved, this page will show all the author info]
![image](https://user-images.githubusercontent.com/38216903/58281434-05d5c300-7d58-11e9-9f11-23a9178da7a9.png)

Support topic page:
![image](https://user-images.githubusercontent.com/38216903/58281389-ec347b80-7d57-11e9-9976-5f1595e04654.png)

User profile page:
![image](https://user-images.githubusercontent.com/38216903/58281530-37e72500-7d58-11e9-8fef-04906313946d.png)

rekhaswaminathan commented on ApplensBackend/Services/SelfHelpContentService/SelfHelpContentService.cs:
```diff
@@ -0,0 +1,212 @@
+            }
+            catch (Exception ex)
+            {
+                throw ex;
```
should we also log the exception message and pesId supportTopicId ?

cindylovescoding replied:
Unfortunately we don't have log service in Applens backend currently. I will add that once we enable app insights logging for applens backend in the future.

rekhaswaminathan commented on ApplensBackend/Services/SelfHelpContentService/SelfHelpContentService.cs:
```diff
@@ -0,0 +1,212 @@
+                }
+                catch (Exception ex)
+                {
+                    // Ignore exception for now
```
Same here for logging

rekhaswaminathan commented on ApplensBackend/Services/GraphClientService/GraphTokenService.cs:
```diff
@@ -0,0 +1,94 @@
+                    DateTime invocationEndTime = DateTime.UtcNow;
+                    long latencyInMs = Convert.ToInt64((invocationEndTime - invocationStartTime).TotalMilliseconds);
+
+                    // TODO : Log an Event
```
are you planning to do this now or later?

rekhaswaminathan commented on ApplensBackend/Services/DiagnosticRoleClientService/DiagnosticRoleClient.cs:
```diff
@@ -114,6 +113,7 @@ public async Task<HttpResponseMessage> Execute(string method, string path, strin
                     {
                         var requestMessage = new HttpRequestMessage(HttpMethod.Post, "api/invoke");
                         requestMessage.Headers.Add("x-ms-path-query", path);
+                        requestMessage.Headers.Add("x-ms-internal-client", internalClient.ToString());
```
I thought the UI code is already sending this header. Is this something different ?

cindylovescoding replied:
So the UI code is sending the header info to applens backend controller, however this flag is never used and routed to diag role previously. This flag should also be retrieved from the header and send to the backend.

ShekharGupta1988 commented:
@cindylovescoding  - This is looking so professional compared to current applens. (lol). Great work !!

ShekharGupta1988 commented on ApplensBackend/Services/SupportTopicService/SupportTopicService.cs:
```diff
@@ -0,0 +1,69 @@
+        private IKustoQueryService _kustoQueryService;
+
+        private string _supportTopicsQuery = @"
+        cluster('azsupport').database('AzureSupportability').ActiveSupportTopicTree
```
@nmallick1 - just FYI - We can probably use this kusto query in the ASC code also and remove support topic tree hardcoding.

ShekharGupta1988 commented on ApplensBackend/Services/SelfHelpContentService/SelfHelpContentService.cs:
```diff
@@ -0,0 +1,212 @@
+                            if (!fileData.name.ToString().Contains("-scoping-"))
+                            {
+                                string fileDataUrl = fileData.download_url;
+                                tasks.Add(Task.Run(() => GetFileContent(fileDataUrl)));
```
Why do we need to wrap the Task in Task.Run()
we can simply do tasks.Add(GetFileContent(fileDataUrl);

and then await all?

dagrooms52 replied:
This is confusing because we don't use the "Async" suffix on async methods - what do you think about adding that to the code standards?

ShekharGupta1988 replied:
agreed. adding Async in the name helps.

cindylovescoding commented on AngularApp/projects/applens/src/app/modules/dashboard/dashboard/dashboard.component.ts:
```diff
@@ -53,12 +64,57 @@ export class DashboardComponent implements OnDestroy {
+  }
+
+  navigateToUserPage() {
+    this.navigateTo(`users/xipeng`);
```
This is going to be fixed in my updated PR.

dagrooms52 commented on ApplensBackend/Services/GraphClientService/GraphClientService.cs:
```diff
@@ -0,0 +1,186 @@
+            }
+            catch (Exception ex)
+            {
+                throw ex;
```
This try/catch should be removed, it's not doing anything

cindylovescoding replied:
I have removed all the try-catch clause.

dagrooms52 commented on ApplensBackend/Services/GraphClientService/GraphClientService.cs:
```diff
@@ -0,0 +1,186 @@
+            }
+            catch (Exception ex)
+            {
+                throw ex;
```
Remove try/catch

ShekharGupta1988 commented on ApplensBackend/Services/SelfHelpContentService/SelfHelpContentService.cs:
```diff
@@ -0,0 +1,212 @@
+                                        selfHelpStr = fileContent;
+                                    }
+
+                                    selfHelpMapping.AddOrUpdate(supportTopicIds, fileContent, (key, oldvalue) => fileContent);
```
If possible, consider using TryAddOrUpdate to consume any possible exceptions

dagrooms52 commented on ApplensBackend/Services/GraphClientService/GraphClientService.cs:
```diff
@@ -0,0 +1,186 @@
+
+            try
+            {
+                var tasks = new List<Task>();
```
This looks like an unused variable

dagrooms52 commented on ApplensBackend/Services/GraphClientService/GraphClientService.cs:
```diff
@@ -0,0 +1,186 @@
+            }
+            catch (Exception ex)
+            {
+                throw ex;
```
try/catch doesn't do anything

dagrooms52 commented on ApplensBackend/Services/GraphClientService/GraphClientService.cs:
```diff
@@ -0,0 +1,186 @@
+
+            try
+            {
+                var tasks = new List<Task>();
```
tasks is unused

ShekharGupta1988 commented on ApplensBackend/Services/GraphClientService/GraphTokenService.cs:
```diff
@@ -0,0 +1,94 @@
+
+    public class GraphTokenService : IGraphTokenService
+    {
+        private AuthenticationContext _authContext;
```
We should extract the logic in a base token class and then just derive the kusto token class as well as Graph token class.
Rekha recently did that in backend also.

rekhaswaminathan commented:
The new UI looks great @cindylovescoding!!
A small feedback:
Is it possible to reduce the white space on the left so that detectors with longer names wont flow to next line 
![image](https://user-images.githubusercontent.com/46545195/58291562-bbae0b00-7d72-11e9-8ed4-69db5818c879.png)

rekhaswaminathan commented on ApplensBackend/Services/GraphClientService/GraphClientService.cs:
```diff
@@ -0,0 +1,186 @@
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
```
I found that lot of these imports are not being used, if you have time remove the unused references.

cindylovescoding commented:
Added different resource images for function, linux and ase:
![image](https://user-images.githubusercontent.com/38216903/58298097-0d648e80-7d8f-11e9-9daf-d29d5e58544b.png)
![image](https://user-images.githubusercontent.com/38216903/58298125-279e6c80-7d8f-11e9-8b69-b77158b22789.png)

Added observer link for sites and linux in resource summary modal:
![image](https://user-images.githubusercontent.com/38216903/58298158-47ce2b80-7d8f-11e9-806a-b3afea16b1fd.png)

puneetg1983 commented:
>         "autoprefixer": "^8.4.1",

I don't believe changes to this file need to be checked in. We should update the .gitignore so git ignores this files on commits

---
Refers to: AngularApp/package-lock.json:29 in 9b52383. [](commit_id = 9b5238318b933fc09fa957216d1d2a66670aa14a, deletion_comment = True)

puneetg1983 commented on AngularApp/projects/applens/src/app/modules/dashboard/category-page/category-page.component.ts:
```diff
@@ -0,0 +1,200 @@
+                    this.filterdDetectors.forEach((detector) => {
+                        this._supportTopicService.getCategoryImage(detector.name).subscribe((iconString) => {
+                            let onClick = () => {
+                                this.navigateTo(`../../detectors/${detector.id}`);
```
A check is needed here to check detector.type === DetectorType.Analysis and then the route should be updated to ../../analysis/detectorId

cindylovescoding replied:
So Puneet I think for now all the analysises (DetectorType.Analysis) are tagged with uncategorized right? Which means it won't be showed up in the "By category" page. 

In "By Support topic" page, I have added the analysis router. But I will still add the analysis type check here in case we decide to tag the analysis with some Categories in the future.

puneetg1983 commented on AngularApp/projects/applens/src/app/modules/dashboard/services/applens-support-topic.service.ts:
```diff
@@ -0,0 +1,220 @@
+
+@Injectable()
+export class ApplensSupportTopicService {
+    private categoryKeywordsImagePathMapping = [
```
Nice thought and effort :-)

puneetg1983 commented on AngularApp/projects/applens/src/app/modules/dashboard/services/applens-support-topic.service.ts:
```diff
@@ -0,0 +1,220 @@
+                    detector.supportTopicList.findIndex(supportTopic => supportTopic.id === supportTopicId) >= 0);
+
+                if (matchingDetector) {
+                    detectorPath = `/detectors/${matchingDetector.id}`;
```
Same comment here. The path should be returned after checking detector.type and if DetectorType.Analysis then path should be updated accordingly

puneetg1983 commented on ApplensBackend/appsettings.json:
```diff
@@ -51,5 +51,16 @@
+    "ClientId": "",
+    "AppKey": ""
+  },
+  "SelfHelpContent": {
```
Just curious if this is a different repo than datascripts repo?

cindylovescoding replied:
Yes it is under a different repo: https://github.com/Azure/SelfHelpContent/tree/master/articles/microsoft.web

puneetg1983 commented on ApplensBackend/Models/SupportTopic.cs:
```diff
@@ -0,0 +1,23 @@
+        public string ProductName { get; set; }
+
+
+        public string SupportTopicL2Name { get; set; }
```
>public [](start = 8, length = 6)

nit: extra line above

puneetg1983 commented:
git is reporting some conflicting changes. Just ensure to resolve before merge :)
